map/local_ads_manager.cpp

```diff
@@ -0,0 +1,350 @@
+{
+  auto connectionStatus = GetPlatform().ConnectionStatus();
+  if (connectionStatus == Platform::EConnectionType::CONNECTION_NONE ||
+      df::GetZoomLevel(screen.GetScale()) <= scales::GetUpperWorldScale())
+  {
+    return;
+  }
+
+  std::vector<std::string> requestedCampaigns;
+  auto mwms = m_getMwmsByRectFn(screen.ClipRect());
+  if (mwms.empty())
+    return;
+
+  // Request local ads campaigns.
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    for (auto const & mwm : mwms)
+    {
+      auto info = mwm.GetInfo();
+      if (!info)
+        continue;
+      std::string const & mwmName = info->GetCountryName();
+      auto campaignIt = m_campaigns.find(mwmName);
+      if (campaignIt == m_campaigns.end())
+      {
+        requestedCampaigns.push_back(mwmName);
+        continue;
+      }
+
+      // If a campaign has not been requested from server this session.
+      if (!campaignIt->second)
+      {
+        auto const it = m_expiration.find(mwmName);
+        bool needUpdateByTimeout = (connectionStatus == Platform::EConnectionType::CONNECTION_WIFI);
+        if (!needUpdateByTimeout && it != m_expiration.end())
+        {
+          auto const currentTime = std::chrono::steady_clock::now();
+          needUpdateByTimeout = (currentTime - it->second) > kWWanUpdateTimeout;
+        }
+
+        if (needUpdateByTimeout || it == m_expiration.end())
+          requestedCampaigns.push_back(mwmName);
+      }
+    }
+
+    if (!requestedCampaigns.empty())
+    {
+      m_requestedCampaigns.reserve(m_requestedCampaigns.size() + requestedCampaigns.size());
+      for (auto const & campaign : requestedCampaigns)
+        m_requestedCampaigns.push_back(std::make_pair(m_getMwmIdByNameFn(campaign), RequestType::Download));
+      m_condition.notify_one();
+    }
+  }
+}
+
+void LocalAdsManager::ThreadRoutine()
+{
+  std::string const campaignFile = GetPath(kCampaignFile);
+  std::string const expirationFile = GetPath(kExpirationFile);
+
+  // Read persistence data (expiration file must be read the first).
+  ReadExpirationFile(expirationFile);
+  ReadCampaignFile(campaignFile);
+
+  std::vector<Request> campaignMwms;
+  while (WaitForRequest(campaignMwms))
+  {
+    try
+    {
+      FilesContainerW campaignsContainer(campaignFile, FileWriter::OP_WRITE_EXISTING);
+      FilesContainerW expirationContainer(expirationFile, FileWriter::OP_WRITE_EXISTING);
+      for (auto const & mwm : campaignMwms)
+      {
+        if (!mwm.first.IsAlive())
+          continue;
+
+        std::string const countryName = mwm.first.GetInfo()->GetCountryName();
+        if (mwm.second == RequestType::Download)
+        {
+          // Download campaign data from server.
+          std::vector<uint8_t> rawData = DownloadCampaign(mwm.first);
+          if (rawData.empty())
+            continue;
+
+          // Save data persistently.
+          campaignsContainer.Write(rawData, countryName);
+          auto ts = std::chrono::steady_clock::now();
```

Comment:
Sorry, we can't use steady_clock. It should be used for measuring time intervals during the application lifetime and should not be used for storing timestamps.